cosmology/StructureFormationManager.h

```diff
@@ -0,0 +1,597 @@
+
+        // Since the particles have moved spatially update them to correct processors
+        pc->update();
+
+        bool isFirstRepartition = false;
+        std::shared_ptr<FieldContainer_t> fc    = this->fcontainer_m;
+        if (this->loadbalancer_m->balance(this->totalP_m, this->it_m)) {
+                auto* mesh = &fc->getRho().get_mesh();
+                auto* FL = &fc->getFL();
+                this->loadbalancer_m->repartition(FL, mesh, isFirstRepartition);
+                printf("first repartition works \n");
+        }
+
+        
+        mes << "Assignment of positions and velocities done." << endl;
+
+    }
+
+    /**
+     * @brief Read particle data from a file and assign to the domain.
+     */
+    void readParticlesDomain() {
+        Inform mes("Reading Particles");
+
+        this->fcontainer_m->getRho() = 0.0;
+
+        // Load Balancer Initialisation
+        auto *mesh = &this->fcontainer_m->getMesh();
+        auto *FL = &this->fcontainer_m->getFL();
+        if ((this->lbt_m != 1.0) && (ippl::Comm->size() > 1)) {
+            mes << "Starting first repartition" << endl;
+            this->isFirstRepartition_m           = true;
+            this->loadbalancer_m->initializeORB(FL, mesh);
+            this->loadbalancer_m->repartition(FL, mesh, this->isFirstRepartition_m);
+        }
+
+        static IpplTimings::TimerRef ReadingTimer = IpplTimings::getTimer("readData");
+        IpplTimings::startTimer(ReadingTimer);
+
+        
+        // Check if the file is opened successfully
+        std::ifstream file(this->folder + "Data.csv");
+        if (!file.is_open()) {
+            std::cerr << "Error opening IC file!" << std::endl;
+        }
+        
+        // Vector to store data read from the CSV file
+        std::vector<std::vector<double>> ParticlePositions;
+        std::vector<std::vector<double>> ParticleVelocities;
+
+        // Boundaries of Particle Positions
+        const ippl::NDIndex<Dim>& ldom = FL->getLocalNDIndex(); // local domain in coordinates [256:511:1] [0:511:1] [0:511:1]
+        Vector_t<double, Dim> Min;
+        Vector_t<double, Dim> Max;
+        for(unsigned int i = 0; i<Dim; ++i){
+            Min[i] = this->rmax_m[i] * ldom[i].first() / this->nr_m[i];
+            Max[i] = this->rmax_m[i] * (ldom[i].last()+1) / this->nr_m[i];
+
+        }
+
+
+        // Read the file line by line
+        std::string line;
+        while (std::getline(file, line)) {
+            // New Line has begun
+            std::stringstream ss(line);
+            std::string cell;
+            int j = 0; // column number
+            std::vector<double> PosRow;
+            std::vector<double> VelRow;
+            bool inDomain = true;
+            while (inDomain==true && j < 6 && std::getline(ss, cell, ',')) {
+                if (j < 3){
+                    double Pos = std::stod(cell);
+                    // Special case where particle lies on the edge
```

Comment:
Maybe add a few sentences summarizing what is being done in this special case when a particle lies on an edge.
From Blanca's thesis:
"Furthermore, it was observed that the sending process of the particles is unstable in the case where a particle lies precisely at the boundary of the domain of two processors. For those cases, a small perturbation (0.01 %) of the particle positions was included when reading in the position. This issue has been detected since a few particles have been double-counted which throws an error since the total number of particles is not conserved."